test/test_goal_queueing.cpp

```diff
@@ -0,0 +1,289 @@
+		sleep_cv.wait(ulk, 
+				[this](){return !execution;}
+			); 
+	}
+
+	void resetFlags() {
+		got_goal = false;
+		goal_preempted = false;
+		next_goal_available = false;
+		received_goal = nullptr;
+		
+		// Signal flags
+		execution = false;
+		finish_executing = false;
+		resume_executing = false;
+		execute_done = false;
+	}
+
+	// Flags for assertions
+	bool got_goal;
+	bool goal_preempted;
+	bool next_goal_available;
+	move_base_msgs::MoveBaseGoalConstPtr received_goal;
+
+	// Signaling variables
+	bool execution;
+	bool finish_executing;
+	bool resume_executing;
+	bool execute_done;
+	std::mutex sleep_lock;
+	std::mutex update_lock;
+	std::mutex execute_lock;
+	std::mutex execute_done_lock;
+	std::condition_variable sleep_cv;
+	std::condition_variable execute_cv;
+	std::condition_variable execute_done_cv;
+
+	std::unique_ptr<actionlib::QueuedActionServer<move_base_msgs::MoveBaseAction>> qserv;
+	std::unique_ptr<actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction>> cli;
+};
+
+
+TEST_F(GoalQueueSuite, addGoalWhileExecuting) { 
+	move_base_msgs::MoveBaseGoal goal; 
+
+	// First goal
+	goal.target_pose.pose.position.x = 3.0;
+	cli->sendGoal(goal);
+	ros::spinOnce(); 
+	sleepExecuting();
+	EXPECT_TRUE(got_goal);
+	EXPECT_FALSE(goal_preempted);
+	EXPECT_FALSE(next_goal_available);
+	EXPECT_EQ(3.0, received_goal->target_pose.pose.position.x);
+
+	// Second goal
+	goal.target_pose.pose.position.x = 7.0;
+	cli->sendGoal(goal);
+	ros::spinOnce(); 
+
+	resumeExecuting();
+	updateExecuting();
+	sleepExecuting();
+ 	EXPECT_TRUE(next_goal_available); 
+	
+	// Cancel the last goal sent
+	cli->cancelGoal(); 
+	ros::Duration(1.0).sleep(); 
+	ros::spinOnce(); 
+	EXPECT_EQ(3.0, received_goal->target_pose.pose.position.x); // Making sure we are in the first goal thread
+ 	finishExecuting(); // Finish 1st goal
+	ros::Duration(0.5).sleep(); 
+
+	EXPECT_TRUE(goal_preempted); 
+ 	finishExecuting(); // Finish 2nd (canceled) goal
+	
+	// New goal
+	goal.target_pose.pose.position.x = 13.0;
+	cli->sendGoal(goal); 
+	ros::spinOnce(); 
+	sleepExecuting();
+	EXPECT_TRUE(got_goal);
+	EXPECT_FALSE(goal_preempted);
+	EXPECT_EQ(13.0, received_goal->target_pose.pose.position.x);  
+	finishExecuting(); // Finish new goal
+
+// 	- if another goal is received add it to the queue (DONE) 
+// 	- if the queue full, set the next goal as preempted - explicitly called! - so cancelling the next goal and adding new to the queue
+// 	- start executing the new goal in queue (after the current)
+}
+
+TEST_F(GoalQueueSuite, stopAfterCancel) {
+	move_base_msgs::MoveBaseGoal goal; 
+
+	// One goal -> Cancel request -> Stop
+	goal.target_pose.pose.position.x = 3.0;
+	cli->sendGoal(goal);
+	ros::spinOnce(); 
+	sleepExecuting();
+	EXPECT_TRUE(got_goal);
+	EXPECT_EQ(3.0, received_goal->target_pose.pose.position.x);
+
+	cli->cancelGoal();
+	ros::Duration(1.0).sleep();
+	ros::spinOnce(); 
+	resumeExecuting();
+	updateExecuting();
+	sleepExecuting();
+	EXPECT_TRUE(goal_preempted);	
```

Comment:
I tested this 10 times, this is the only line that failed (failed 3 times). Can you put a 0.5 second sleep before it as a quick fix?

Reply:
Weird, never failed in my case. Yes, sure.